server/src/main/java/io/deephaven/server/hierarchicaltable/HierarchicalTableViewSubscription.java

```diff
@@ -0,0 +1,496 @@
+        listener.onNext(streamGenerator.getSubView(subscriptionOptions, true, rows, false, rows, columns));
+
+        // 6. Let the caller know what the expanded size was
+        return expandedSize;
+    }
+
+    public void setViewport(
+            @Nullable final BitSet viewportColumns,
+            @Nullable final RowSet viewportRows,
+            final boolean reverseViewport) {
+
+        if (state != State.Active) {
+            return;
+        }
+
+        if (viewportColumns != null) {
+            if (viewportColumns.length() > view.getHierarchicalTable().getSnapshotDefinition().numColumns()) {
+                throw GrpcUtil.statusRuntimeException(Code.INVALID_ARGUMENT, String.format(
+                        "Requested columns out of range: length=%d, available length=%d",
+                        viewportColumns.length(), view.getHierarchicalTable().getSnapshotDefinition().numColumns()));
+            }
+        }
+        if (viewportRows != null) {
+            if (!viewportRows.isContiguous()) {
+                throw GrpcUtil.statusRuntimeException(Code.INVALID_ARGUMENT,
+                        "HierarchicalTableView subscriptions only support contiguous viewports");
+            }
+            if (viewportRows.size() > LARGEST_POOLED_CHUNK_CAPACITY) {
+                throw GrpcUtil.statusRuntimeException(Code.INVALID_ARGUMENT, String.format(
+                        "HierarchicalTableView subscriptions only support viewport size up to %d rows, requested %d rows",
+                        LARGEST_POOLED_CHUNK_CAPACITY, viewportRows.size()));
+            }
+        }
+        if (reverseViewport) {
+            throw GrpcUtil.statusRuntimeException(Code.INVALID_ARGUMENT,
+                    "HierarchicalTableView subscriptions do not support reverse viewports");
+        }
+
+        final BitSet newColumns = viewportColumns == null ? null : (BitSet) viewportColumns.clone();
+        final RowSet newRows = viewportRows == null ? null : viewportRows.copy();
+        final long currentTimeNanos = System.nanoTime();
+        synchronized (schedulingLock) {
+            if (state != State.Active) {
+                return;
+            }
+            if (newColumns != null) {
+                this.pendingColumns = newColumns;
+            }
+            if (newRows != null) {
+                try (final SafeCloseable ignored = this.pendingRows) {
+                    this.pendingRows = newRows;
+                }
+            }
+            if (newColumns != null || newRows != null) {
+                scheduleImmediately(currentTimeNanos);
+            }
+        }
+    }
+
+    private void scheduleImmediately(final long currentTimeNanos) {
+        Assert.holdsLock(schedulingLock, "schedulingLock");
+        if (currentTimeNanos < scheduledTimeNanos) {
+            scheduledTimeNanos = currentTimeNanos;
+            scheduler.runImmediately(propagationJob);
```

Comment:
I *think* I've addressed the state keeping necessary to avoid scheduling more than one at a time, except one overlapping an actively running job. I thought it was friendlier to other users of the scheduler to give up my slot and wait to be executed again. If you agree that I've met my goal of only having at most two jobs in the scheduler (one running, one pending), I think I'd prefer to leave it as is.